frontend/test/pytest/test_jax_integration.py

```diff
@@ -0,0 +1,253 @@
+            qml.RX(jnp.pi * x[0], wires=0)
+            qml.RY(x[1] ** 2, wires=0)
+            qml.RX(y[1] * x[2], wires=0)
+            return qml.probs(wires=0), qml.expval(qml.PauliZ(0))
+
+        @partial(jax.grad, argnums=[0, 1])
+        def cost_fn(x, y):
+            result = circuit(x, y)
+            return jnp.sum(jnp.cos(result[0]) ** 2) + jnp.sin(result[1]) ** 2
+
+        result = cost_fn(jnp.array([0.1, 0.2, 0.3]), jnp.array([0.1, 0.2]))
+        assert len(result) == 2
+        assert jnp.allclose(result[0], jnp.array([-0.6027582, -0.00797699, -0.03839391]))
+        assert jnp.allclose(result[1], jnp.array([0.0, -0.05759086]))
+
+    def test_without_precompilation(self):
+        """Test a function without type hints (pre-compilation) using jax.grad on top of qjit."""
+
+        @qjit
+        @qml.qnode(qml.device("lightning.qubit", wires=2))
+        def circuit(x, y):
+            qml.RX(jnp.pi * x[0], wires=0)
+            qml.RY(x[1] ** 2, wires=0)
+            qml.RX(y[1] * x[2], wires=0)
+            return qml.probs(wires=0), qml.expval(qml.PauliZ(0))
+
+        @partial(jax.grad, argnums=[0, 1])
+        def cost_fn(x, y):
+            result = circuit(x, y)
+            return jnp.sum(jnp.cos(result[0]) ** 2) + jnp.sin(result[1]) ** 2
+
+        result = cost_fn(jnp.array([0.1, 0.2, 0.3]), jnp.array([0.1, 0.2]))
+        assert len(result) == 2
+        assert jnp.allclose(result[0], jnp.array([-0.6027582, -0.00797699, -0.03839391]))
+        assert jnp.allclose(result[1], jnp.array([0.0, -0.05759086]))
+
+    def test_non_differentiable_arguments(self):
+        """Test a circuit with non-differentiable arguments using jax.grad on top of qjit."""
+
+        @qjit
+        @qml.qnode(qml.device("lightning.qubit", wires=2))
+        def circuit(x: jax.ShapedArray((3,), dtype=float), y: int):
+            qml.RX(jnp.pi * x[0], wires=0)
+            qml.RY(x[1] ** 2, wires=0)
+            qml.RX(y * x[2], wires=0)
+            return qml.probs(wires=0), qml.expval(qml.PauliZ(0))
+
+        @partial(jax.grad)
+        def cost_fn(x, y):
+            result = circuit(x, y)
+            return jnp.sum(jnp.cos(result[0]) ** 2) + jnp.sin(result[1]) ** 2
+
+        result = cost_fn(jnp.array([0.1, 0.2, 0.3]), 3)
+        assert jnp.allclose(result, jnp.array([-1.34682245, -0.00432698, -1.2855136]))
+
+    def test_multiple_calls(self):
+        """Test a jax.jit function which repeatedly calls a qjit function."""
+
+        @qjit
+        @qml.qnode(qml.device("lightning.qubit", wires=1))
+        def circuit(x):
+            qml.RY(x, wires=0)
+            return jnp.asarray(measure(0), dtype=float)
+
+        @jax.grad
+        def cost_fn(x, y):
+            m1 = circuit(x)
+            m2 = circuit(y)
+            return m1 + m2
+
+        result1 = cost_fn(0.0, 0.0)
+        result2 = cost_fn(0.0, jnp.pi)
+        assert jnp.allclose(result1, 0.0)
+        assert jnp.allclose(result2, 0.0)
+
```

Comment:
@dime10 it might be worth also adding tests for:

- `jax.jacobian` with a vector valued QNode
- `jax.vmap`, assuming it is working?

Reply:
Good idea, will add!

Reply:
https://github.com/PennyLaneAI/catalyst/pull/96/commits/881669a5662d3f9516e088fc921fb9311298449f